File: 21558 - Nithin/week 10/152-maximum-product-subarray/152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        if(n == 1){
            return nums[0];
        }
        int max_so_far = nums[0], max_ending_here = nums[0],min_ending_here = nums[0];
        for(int i = 1;i < n;i++){
            int temp = max({nums[i],nums[i]*max_ending_here,nums[i]*min_ending_here});
            min_ending_here = min({nums[i],nums[i]*max_ending_here,nums[i]*min_ending_here});
            max_ending_here = temp;
            max_so_far = max(max_ending_here, max_so_far);
        }
        return max_so_far;
    }
};

File: 21561 - Suvam Tamang/Week 10/Maximum Product Subarray/152. Maximum Product Subarray.java
class Solution {
    public int maxProduct(int[] nums) {

        // // This is the brute force method of solving the given problem
        // int n = nums.length;
        // int max = nums[0];

        // for(int i=0; i<n; i++) {
        //   int temp = 1;
        //   for(int j=i; j<n; j++) {
        //     temp *= nums[j];
        //     max = Math.max(max, temp);
        //   }
        // }

        // return max;

        // Using Dynamic Programming

        int n = nums.length;
        int local_max = nums[0];
        int local_min = nums[0];
        int global_max = nums[0];

        for(int i=1; i<n; i++) {
          int temp = local_max;
          local_max = Math.max(temp * nums[i], Math.max(local_min * nums[i], nums[i]));
          local_min = Math.min(temp * nums[i], Math.min(local_min * nums[i], nums[i]));

          global_max = Math.max(local_max, global_max);
        }
        return global_max;
    }
}
